sdk/src/Core/RegionEndpoint/RegionEndpointProviderV3.cs

```diff
@@ -364,34 +374,51 @@ public void Add(string serviceName, RegionEndpoint.Endpoint endpoint, HashSet<st
             {
                 if (variants == null || variants.Count == 0)
                 {
-                    _serviceMap.Add(serviceName, endpoint);
+                    lock (_serviceMapLock)
+                    {
+                        _serviceMap.Add(serviceName, endpoint);
+                    }
                 }
                 else
                 {
-                    if (!_variantMap.ContainsKey(serviceName) || _variantMap[serviceName] == null)
```

Comment:
If '_variantMap[serviceName] == null' we will get exception of KeyAlreadyExist. That why I spitted the if statement to two.